server/src/main/java/org/elasticsearch/indices/recovery/RemoteRecoveryTargetHandler.java

```diff
@@ -202,12 +203,17 @@ public void writeFileChunk(StoreFileMetadata fileMetadata, long position, Releas
         final RecoveryFileChunkRequest request = new RecoveryFileChunkRequest(
             recoveryId, requestSeqNo, shardId, fileMetadata, position, content, lastChunk, totalTranslogOps, throttleTimeInNanos);
         final Writeable.Reader<TransportResponse.Empty> reader = in -> TransportResponse.Empty.INSTANCE;
-        executeRetryableAction(
-            action,
-            request,
-            fileChunkRequestOptions,
-            ActionListener.runBefore(listener.map(r -> null), request::decRef),
-            reader);
+
+        // Fork the actual sending onto a separate thread so we can send them concurrently even if CPU-bound (e.g. using compression).
+        // The AsyncIOProcessor and MultiFileWriter both concentrate their work onto fewer threads if possible, but once we have
+        // chunks to send we want to increase parallelism again.
+        threadPool.generic().execute(ActionRunnable.wrap(listener, l ->
```

Comment:
Do you know why compression is so expensive (i.e. which part of it?). I wonder if we could get an improvement out of increasing the chunk size if most of the cost is the per compression start/end overhead (maybe there's something to improve there as well).

Reply:
I think it's just that `DEFLATE` is expensive, especially if the data is already deflated (as it is for stored fields which make up the bulk of this index). `pv _1t.fdt | gzip > /dev/null` indicates my machine can only sustain 25MiBps on a single thread, which is about the performance we see with no parallelism.

Reply:
Right now I remember us benchmarking DEFLATE before => nothing we can do but parallelize I guess :)